### ISSUE TITLE
Question about parameter k of SLIM

### ISSUE THREAD

**Issue by alexandre-termier** [scikit-mine/scikit-mine#121]
#### Describe the issue linked to the documentation

The parameter k is described as the number of non-singleton itemsets to mine. 
While this is correct in the code, it is not very clear for the user of what is its interest, making the setting of this parameter tricky. 

#### Suggest a potential alternative/fix

In practice, the real use of k here is to control the number of iterations of SLIM. 
Putting higher values of k helps SLIM to test out more candidates, and thus to further refine the discovered itemsets. It will also allow SLIM to make more combinations of smaller itemsets to try and make bigger itemsets. 

In some ways, tol is also just a way to reduce execution time (at the cost of the quality of results). tol is not clean either as the 0.5 value has no real meaning (it depends too much on the dataset at hand).

I see two possible solutions:
  - small change: modify the behavior of k to make it a number of iterations, and state in doc that more iterations = possibly a better code table
  - more involved change: remove k and tol and replace them by a time budget, that could be set by default at 30 seconds. This requires more work, but has the advantage of having a clear semantics for the user, and also to let the algorithm perform unhindered until the time out.

**Comment by remiadon** [comment]
Hi there,

## Some context 
the k parameter is here to control the "dimension" of the outputted codetable. In discovery mode, we just control the number of patterns returns via `.discover()`, but if we use it as a feature extractor (for machine learning pipelines for example), we have to control the number of patterns themselves (eq to the dimension of the resulting matrix return by `.transform()`), and not the number of iteration.
The discovery mode is the one you are using, documented [here](https://scikit-mine.github.io/scikit-mine/reference/itemsets.html#slim)
The feature extractor is documented [here](https://scikit-mine.github.io/scikit-mine/reference/feature_extraction.html#slimvectorizer)

## To answer your question
### k parameter
I agree we could give control over the number of iteration. For compatibility with sklearn however, I think this param should be renamed to `n_iter` (see the sklearn [KMeans](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html) for example.
The only disadvantage is that we cannot guarantee a number of patterns to be returned to our user (or we could pass this `k` parameter to `.discover` just when the codetable is exposed)

 In any case we should keep the `k` parameter as it is for the SLIMVectorizer class, which you are not using :)

### Time budget
IMO setting a time budget is not a good methodology if we strive for reproducible research. CPU time and memory consumption are metrics we want to optimise when setting specific parameters, not the other way around. In addition to that, this would make possible running the same method with the same parameters (the time budget) and obtaining different models on different physical environments (eg. machines). 

If users come to us for debugging, we will struggle to address the results they share with regard to a specific parameter setting.

**Comment by thomasbtnfr** [comment]
A `max_time` parameter has now been implemented for SLIM.
